cleverhans/train.py

```diff
@@ -104,78 +106,122 @@ def train(sess, loss, x_train, y_train,
-            while len(index_shuf) % batch_size != 0:
-                index_shuf.append(rng.randint(len(x_train)))
-            nb_batches = len(index_shuf) // batch_size
-            rng.shuffle(index_shuf)
-            # Shuffling here versus inside the loop doesn't seem to affect
-            # timing very much, but shuffling here makes the code slightly
-            # easier to read
-            x_train_shuffled = x_train[index_shuf]
-            y_train_shuffled = y_train[index_shuf]
-
-            prev = time.time()
-            for batch in range(nb_batches):
-
-                # Compute batch start and end indices
-                start = batch * batch_size
-                end = (batch + 1) * batch_size
-                # start, end = batch_indices(
-                #    batch, len(x_train), args.batch_size)
-
-                # Perform one training step
-                feed_dict = {}
-                diff = end - start
-                assert diff == batch_size
-                stride = diff // num_devices
-                for dev_idx in xrange(num_devices):
-                    cur_start = start + dev_idx * stride
-                    cur_end = start + (dev_idx + 1) * stride
-                    feed_dict[xs[dev_idx]
-                              ] = x_train_shuffled[cur_start:cur_end]
-                    feed_dict[ys[dev_idx]
-                              ] = y_train_shuffled[cur_start:cur_end]
-                if cur_end != end:
-                    msg = ("batch_size (%d) must be a multiple of num_devices "
-                           "(%d).\nCUDA_VISIBLE_DEVICES: %s"
-                           "\ndevices: %s")
-                    args = (batch_size, num_devices,
-                            os.environ['CUDA_VISIBLE_DEVICES'],
-                            str(devices))
-                    raise ValueError(msg % args)
-                if feed is not None:
-                    feed_dict.update(feed)
-                sess.run(train_step, feed_dict=feed_dict)
-            assert end == len(index_shuf)  # Check that all examples were used
-            cur = time.time()
-            _logger.info("Epoch " + str(epoch) + " took " +
-                         str(cur - prev) + " seconds")
-            if evaluate is not None:
-                evaluate()
+    assert batch_size % num_devices == 0
+    device_batch_size = batch_size // num_devices
+
+    if init_all:
+        sess.run(tf.global_variables_initializer())
+    else:
+        initialize_uninitialized_global_variables(sess)
+
+    # Check whether the hardware is working correctly
+
+    # So far the failure has only been observed with 3 or more GPUs
+    run_canary = num_devices > 2
+    if run_canary:
+        canary_feed_dict = {}
+        for x, y in safe_zip(preprocessed_xs, ys):
+            canary_feed_dict[x] = x_train[:device_batch_size].copy()
+            canary_feed_dict[y] = y_train[:device_batch_size].copy()
+        param_to_test = 0
```

Comment:
Is this set to 0 because the canary only tests the gradient corresponding to the first parameter in the list of gradients returned by the optimizer's `compute_gradient` ?

Reply:
Yes. I added some comments to explain that.

Reply:
Thanks